module/VuFind/src/VuFind/Auth/AlmaDatabase.php

```diff
@@ -0,0 +1,133 @@
+ * @package  Authentication
+ * @author   Michael Birkner <michael.birkner@akwien.at>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:authentication_handlers Wiki
+ */
+class AlmaDatabase extends Database
+{
+    /**
+     * ILS Authenticator
+     *
+     * @var \VuFind\Auth\ILSAuthenticator
+     */
+    protected $authenticator;
+
+    /**
+     * Catalog connection
+     *
+     * @var \VuFind\ILS\Connection
+     */
+    protected $catalog = null;
+
+    /**
+     * Alma driver
+     *
+     * @var \VuFind\ILS\Driver\Alma
+     */
+    protected $almaDriver = null;
+
+    /**
+     * Alma config
+     *
+     * @var array
+     */
+    protected $almaConfig = null;
+
+    /**
+     * Constructor
+     *
+     * @param \VuFind\ILS\Connection        $connection    The ILS connection
+     * @param \VuFind\Auth\ILSAuthenticator $authenticator The ILS authenticator
+     */
+    public function __construct(
+        \VuFind\ILS\Connection $connection,
+        \VuFind\Auth\ILSAuthenticator $authenticator
+    ) {
+        $this->catalog = $connection;
+        $this->authenticator = $authenticator;
+        $this->almaDriver = $connection->getDriver();
+        $this->almaConfig = $connection->getDriverConfig();
+    }
+
+    /**
+     * Create a new user account in Alma AND in the VuFind Database.
+     *
+     * @param \Zend\Http\PhpEnvironment\Request $request Request object containing
+     *                                                   new account details.
+     *
+     * @return NULL|\VuFind\Db\Row\User New user row.
+     */
+    public function create($request)
+    {
+        // When in privacy mode, don't create an Alma account and delegate
+        // further code execution to the parent.
+        if ($this->getConfig()->Authentication->privacy) {
+            return parent::create($request);
+        }
+
+        // User variable
+        $user = null;
+
+        $params = $this->collectParamsFromRequest($request);
+        $this->validateParams($params);
+
+        // Create user account in Alma
+        $almaAnswer = $this->almaDriver->createAlmaUser($params);
+
+        // Create user account in VuFind user table if Alma gave us an answer
+        if ($almaAnswer !== null) {
+            // If we got this far, we're ready to create the account:
+            $user = $this->createUserFromParams($params);
+
+            // Add the Alma primary ID as cat_id to the VuFind user table
+            $user->cat_id = $almaAnswer->primary_id ?? null;
+
+            // Save the new user to the user table
+            $user->save();
+
+            // Save the credentials to cat_username and cat_password to bypass
+            // the ILS login screen from VuFind
+            $user->saveCredentials($params['username'], $params['password']);
+        } else {
+            throw new AuthException('Error while creating user in Alma');
```

Comment:
The message in an AuthException will be displayed to the end user. It might be worth replacing this with something more generic that can be found in the language files already.

Reply:
I use text that can be translated now.